{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "242706f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import io\n",
    "import matplotlib.pyplot as plt\n",
    "import networkx as nx\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import pyreadr as pyreadr\n",
    "import random\n",
    "import re\n",
    "from scipy import io as sio\n",
    "from scipy import sparse\n",
    "from sh import gunzip\n",
    "from tqdm import tqdm\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5fd85183",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_vcf(path: str) -> 'DataFrame':\n",
    "    '''\n",
    "    Reads vcf file contained in given path and returns as pandas df\n",
    "    *** code from https://gist.github.com/dceoy/99d976a2c01e7f0ba1c813778f9db744 ***\n",
    "    \n",
    "    Parameters:\n",
    "        path (str) -- path to a VCF file\n",
    "        \n",
    "    Returns:\n",
    "        pandas DataFrame containing data from given VCF file\n",
    "    '''\n",
    "    with open(path, 'r') as f:\n",
    "        lines = [l for l in f if not l.startswith('##')]\n",
    "    return pd.read_csv(\n",
    "        io.StringIO(''.join(lines)),\n",
    "        dtype={'#CHROM': str, 'POS': int, 'ID': str, 'REF': str, 'ALT': str,\n",
    "               'QUAL': str, 'FILTER': str, 'INFO': str},\n",
    "        sep='\\t'\n",
    "    ).rename(columns={'#CHROM': 'CHROM'})\n",
    "\n",
    "def process_vcf(path: str, zipped: bool = True, get_annotations: bool = True) -> list:\n",
    "    '''\n",
    "    Unzips and reads VCF file, edits resulting DataFrame to contain only \n",
    "    chromosome, position, and annotation information (if specified)\n",
    "    \n",
    "    Parameters:\n",
    "        path (str) -- path to a VCF file\n",
    "        zipped (bool) -- if true, unzips VCF file [default]\n",
    "        get_annotations (bool) -- if true, extracts and adds annotation data to df\n",
    "        \n",
    "    Returns:\n",
    "        DataFrame representation of VCF file, and list of all annotation types -- if \n",
    "        get_annotations = True\n",
    "    '''\n",
    "    if zipped == True:\n",
    "        gunzip(path)\n",
    "        \n",
    "    vcf = read_vcf(path)\n",
    "    mutations = vcf.loc[:, \"CHROM\":\"ALT\"]\n",
    "    \n",
    "    if get_annotations == True:\n",
    "    \n",
    "        annotations = [0] * len(vcf)\n",
    "        info = vcf[\"INFO\"]\n",
    "        mutation_types = []\n",
    "\n",
    "        for i in range(len(info)):\n",
    "            try:\n",
    "                annotation = re.search(\"Variant_Classification=([A-Z*a-z*0-9*,?'?]+[\\s]*[A-Z*a-z*0-9*,?'?]*)\", info[i]).group(1)\n",
    "                if annotation not in mutation_types:\n",
    "                    mutation_types.append(annotation)\n",
    "                annotations[i] = annotation\n",
    "            except:\n",
    "                annotations[i] = \"NA\"\n",
    "\n",
    "        mutations.loc[:, \"ANNOTATION\"] = annotations\n",
    "\n",
    "        return mutations, mutation_types\n",
    "    \n",
    "    else:\n",
    "        return mutations\n",
    "    \n",
    "def get_mutations_by_bin(mutations: 'DataFrame', bins: 'DataFrame') -> list:\n",
    "    '''\n",
    "    Accepts DataFrame of mutations and DataFrame of scATAC-seq bins and creates\n",
    "    list of the number of mutations per bin\n",
    "    \n",
    "    Parameters:\n",
    "        mutations (df) -- DataFrame containing chromosome and base pair position of mutations\n",
    "        bins (df) -- DataFrame containing chromosome, starting position, and ending position of scATAC-seq bins\n",
    "        \n",
    "    Returns:\n",
    "        list of mutations per bin, length = total # of bins\n",
    "    '''\n",
    "    mutations_by_bin = [0]*len(bins)\n",
    "\n",
    "    for i in tqdm(list(mutations.index.values)):\n",
    "        chrom = mutations.loc[i,'CHROM']\n",
    "        pos = mutations.loc[i, 'POS']\n",
    "        c = f\"chr{chrom}\"\n",
    "        chrom_bins = bins[bins[\"CHROM\"] == c]\n",
    "        try:\n",
    "            bin_index = chrom_bins[(chrom_bins['start'] <= pos) & (chrom_bins['end'] >= pos)].index[0]\n",
    "            mutations_by_bin[bin_index] += 1\n",
    "        except:\n",
    "            continue\n",
    "                \n",
    "    return mutations_by_bin\n",
    "\n",
    "def get_mutation_matrix(processed_vcfs: list, bins: 'DataFrame', filter_by: str = None) -> 'DataFrame':\n",
    "    '''\n",
    "    Accepts list of processed VCF files and creates a mutation df, where each row\n",
    "    contains the number of mutations per bin for each VCF file\n",
    "    \n",
    "    Parameters:\n",
    "        processed_vcfs (list) -- list of processed VCF files output by process_vcf\n",
    "        bins (df) -- DataFrame containing chromosome, starting position, and ending position of scATAC-seq bins\n",
    "        filter_by (str or None) -- annotation type to use in filtering mutations, None is default -- \n",
    "                                   if filter_by = None, no filtering will be performed\n",
    "                                   \n",
    "    Returns:\n",
    "        DataFrame of mutations by bins for each VCF file provided\n",
    "    '''\n",
    "    if filter_by == None:\n",
    "        individual_mutations = []\n",
    "        for v in processed_vcfs:\n",
    "            mutations = get_mutations_by_bin(v, bins)\n",
    "            individual_mutations.append(mutations)\n",
    "        mutation_matrix = pd.DataFrame(individual_mutations)\n",
    "    \n",
    "    else:\n",
    "        annotated_mutations = []\n",
    "        for v in processed_vcfs:\n",
    "            mutations = filter_mutations(v, bins, filter_by)\n",
    "            annotated_mutations.append(mutations)\n",
    "        mutation_matrix = pd.DataFrame(annotated_mutations)\n",
    "        \n",
    "    return mutation_matrix\n",
    "\n",
    "def filter_mutations(annotated_mutations: 'DataFrame', bins: 'DataFrame', annotation: str) -> list:\n",
    "    '''\n",
    "    Accepts DataFrame of mutations and DataFrame of scATAC-seq bins and finds mutations per bin\n",
    "    for specified annotation type\n",
    "    \n",
    "    Parameters:\n",
    "        annotated_mutations (df) -- DataFrame containing chromosome, base pair position, and annotation type of mutations\n",
    "        bins (df) -- DataFrame containing chromosome, starting position, and ending position of scATAC-seq bins\n",
    "        annotation (str) -- annotation type to use in filtering mutations\n",
    "    \n",
    "    Returns:\n",
    "        list of mutations of given annotation type per bin\n",
    "    '''\n",
    "    filtered_mutations = annotated_mutations[annotated_mutations[\"ANNOTATION\"] == annotation]\n",
    "    filtered_mutations_by_bin = get_mutations_by_bin(filtered_mutations, bins)\n",
    "    return filtered_mutations_by_bin\n",
    "\n",
    "def get_cell_scores(mutation_matrix: 'DataFrame | np.ndarray', atac_sparsematrix: np.ndarray, calculate_total: bool = False) -> np.ndarray:\n",
    "    '''\n",
    "    Accepts DataFrame of mutations and sparse matrix from scATAC-sequencing, returns array\n",
    "    of cell scores (total accessible mutations per cell / total accessibility of the cell)\n",
    "    calculated by\n",
    "        1. multiplying mutation array by the ATAC sparse matrix\n",
    "        2. dividing product of #1 by the summed ATAC sparse matrix\n",
    "        \n",
    "    Parameters:\n",
    "        mutation_matrix (DataFrame | np.ndarray) -- matrix of mutations per cell, each rows could be \n",
    "                                                    individuals or mutation types\n",
    "        atac_sparsematrix (np.ndarray) -- binary sparse matrix from scATAC data\n",
    "        calculate_total (bool) -- if True, a final row of the sum of all rows in the matrix will be\n",
    "                                  added to the matrix prior to multiplying and dividing (default = False)\n",
    "        \n",
    "    Returns:\n",
    "        matrix of cell scores, as np.ndarray\n",
    "    '''\n",
    "    if calculate_total == True:\n",
    "        totals = []\n",
    "        for i in range(mutation_matrix.shape[1]):\n",
    "            totals.append(sum(mutation_matrix.loc[:,i]))\n",
    "        mutation_matrix = np.r_[mutation_matrix, [totals]]\n",
    "    \n",
    "    accessible_mutations = mutation_matrix @ atac_sparsematrix\n",
    "    atac_summed = atac_sparsematrix.sum(axis=0)\n",
    "    cell_scores = np.divide(accessible_mutations, atac_summed)\n",
    "    return cell_scores\n",
    "\n",
    "def get_pagerankdicts(cellscore_matrix: 'np.ndarray', atac_cellnames: 'DataFrame') -> list[dict]:\n",
    "    '''\n",
    "    Prepares personalization dictionaries for use in seeding the PageRank algorithm\n",
    "    \n",
    "    Parameters:\n",
    "        cellscore_matrix (np.ndarray) -- numpy matrix of cell scores, where rows are individuals\n",
    "                                         or mutation types\n",
    "        ATAC_cellnames (DataFrame) -- pandas DataFrame of all scATAC-seq cell names contained in graph, \n",
    "                                      should be only column in df\n",
    "                                      \n",
    "    Returns:\n",
    "        list of dictionaries to be used to personalize PageRank\n",
    "    '''\n",
    "    pagerank_dicts = []\n",
    "\n",
    "    for i in range(len(cellscore_matrix)):\n",
    "\n",
    "        row = cellscore_matrix[i]\n",
    "        row_dict = {}\n",
    "\n",
    "        for k in range(row.shape[1]):\n",
    "            row_dict[atac_cellnames.iloc[k]] = row[0,k]\n",
    "\n",
    "        pagerank_dicts.append(row_dict)\n",
    "    \n",
    "    return pagerank_dicts\n",
    "\n",
    "def get_pagerank_csv(G: 'networkx graph obj', pagerank_dicts: list, rna_cellnames: 'DataFrame', path: str, column_names: list[str]) -> None:\n",
    "    '''\n",
    "    Performs personalized PageRank for all dictionaries provided using given graph, and\n",
    "    saves results as a csv file for further analysis and data visualization\n",
    "    \n",
    "    Parameters:\n",
    "        G (networkx graph object) -- networkx graph to be used to perform PageRank\n",
    "        pagerank_dicts (list) -- list of dictionaries to be used to personalize PageRank\n",
    "        rna_cellnames (DataFrame) -- DataFrame containing cell names from scRNA-seq cells found in given graph\n",
    "        path (str) -- path to directory and filename that csv file will be saved to\n",
    "        column_names (list[str]) -- list of names of columns to be used in output csv file\n",
    "        \n",
    "    Returns:\n",
    "        None\n",
    "    '''\n",
    "    csv = pd.DataFrame({\"cell\": rna_cells})\n",
    "    \n",
    "    for i in range(len(pagerank_dicts)):\n",
    "        pr = nx.pagerank(G, personalization = pagerank_dicts[i])\n",
    "        pr = list(pr.values())\n",
    "        pr = pr[0:len(rna_cellnames)]\n",
    "        csv[column_names[i]] = pr\n",
    "    \n",
    "    csv = csv.set_index(\"cell\")\n",
    "    csv.to_csv(path)\n",
    "    \n",
    "def artificially_mutate(bins: 'DataFrame', chromosome: str, bp_start: int, bp_end: int, mutation_max: int, random: bool = False) -> list:\n",
    "    '''\n",
    "    Artificially mutates the scATAC-seq bins for a specific chromosome based on a given bp interval and max number \n",
    "    of mutations (randomized or not) \n",
    "    \n",
    "    Parameters:\n",
    "        bins (df) -- DataFrame containing chromosome, starting position, and ending position of scATAC-seq bins\n",
    "        chromosome (str) -- number or letter corresponding to chromosome to mutate, as string\n",
    "        bp_start (int) -- starting bp for artificial mutations\n",
    "        bp_end (int) -- ending bp for artificial mutations\n",
    "        mutation_max (int) -- maximum number of mutations to be assigned to each bin in range; if random = False, \n",
    "                              all bins will be assigned this number; if random = True, bins will be assigned a \n",
    "                              random value from 1 to this number\n",
    "        random (bool) -- default = False; if True, each bin in the bp range will be assigned a random number of\n",
    "                         mutations\n",
    "        \n",
    "    Returns:\n",
    "        list of artificial mutations per bin, length = total # of bins\n",
    "    '''\n",
    "    mutations_by_bin = [0]*len(ArchR_bins)\n",
    "    \n",
    "    c = f\"chr{chromosome}\"\n",
    "    chrom_bins = bins[bins[\"CHROM\"] == c]\n",
    "\n",
    "    for i in tqdm(range(len(bins))):\n",
    "        start = bins.loc[i, 'start']\n",
    "        end = bins.loc[i, 'end']\n",
    "        if start <= bp_end and end >= bp_start:\n",
    "            if random == True:\n",
    "                integer = random.randint(1, mutation_max)\n",
    "                mutations_by_bin[i] += integer\n",
    "            else:\n",
    "                mutations_by_bin[i] += 1\n",
    "                \n",
    "    return mutations_by_bin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6bc549a4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF files processed and annotated\n",
      "all mutation types: ['lincRNA', 'Missense', 'IGR', 'Intron', \"3'UTR\", 'Silent', \"5'Flank\", 'RNA', 'Nonsense', \"5'UTR\", 'Splice']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 4228/4228 [20:23<00:00,  3.46it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "built mutation matrix -- shape = (1, 6072620)\n",
      "got cell score matrix -- shape = (2, 44514)\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'get_pagerank_dicts' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 61\u001b[0m\n\u001b[1;32m     58\u001b[0m     get_pagerank_csv(G_mpal, pagerank_dicts, rna_cellnames, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m/projectnb/paxlab/randers1/pagerank_all.csv\u001b[39m\u001b[38;5;124m\"\u001b[39m, [\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mall1\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mall2\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mall3\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mall4\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mall5\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mall6\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mall7\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mall8\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mall\u001b[39m\u001b[38;5;124m\"\u001b[39m])\n\u001b[1;32m     59\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPageRank is complete, results can be found in the csv file in the directory indicated\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m---> 61\u001b[0m \u001b[43mmain\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[7], line 54\u001b[0m, in \u001b[0;36mmain\u001b[0;34m()\u001b[0m\n\u001b[1;32m     51\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgot cell score matrix -- shape = \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mcell_scores\u001b[38;5;241m.\u001b[39mshape\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     53\u001b[0m \u001b[38;5;66;03m# create pagerank dictionaries -- all mutation types\u001b[39;00m\n\u001b[0;32m---> 54\u001b[0m pagerank_dicts \u001b[38;5;241m=\u001b[39m \u001b[43mget_pagerank_dicts\u001b[49m(cell_scores, atac_cellnames)\n\u001b[1;32m     55\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgot \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mlen\u001b[39m(pagerank_dicts)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m pagerank dictionaries\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     57\u001b[0m \u001b[38;5;66;03m# get pagerank csv file -- all mutation types\u001b[39;00m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'get_pagerank_dicts' is not defined"
     ]
    }
   ],
   "source": [
    "def main():\n",
    "    \n",
    "    # load scATAC-seq binary sparsematrix here\n",
    "    with open('atac_sparsematrix.pickle', 'rb') as f:\n",
    "        atac_sparsematrix = pickle.load(f)\n",
    "        \n",
    "    # load scATAC-seq bins here\n",
    "    with open('ArchR_bins.pickle', 'rb') as f:\n",
    "        bins = pickle.load(f)\n",
    "    \n",
    "    # load scATAC-seq cell names here\n",
    "    atac_cellnames = pd.read_csv(\"/projectnb/paxlab/agokhale/MPAL/mpal_atac_cells.csv\")\n",
    "    atac_cellnames = atac_cellnames.iloc[:,1]\n",
    "\n",
    "    # load scRNA-seq cell names here\n",
    "    rna_cellnames = pd.read_csv('/projectnb/paxlab/agokhale/MPAL/mpal_rna_cells.csv')\n",
    "    rna_cellnames = rna_cellnames.iloc[:,1]\n",
    "    \n",
    "    # load list of VCF files\n",
    "    vcf_files = [f for f in os.listdir('/projectnb/paxlab/DATA/ICGC/WGS/LAML-KR/') if re.search('.consensus.20160830.somatic.snv_mnv.vcf$', f)]\n",
    "    \n",
    "    # load network\n",
    "    with open(\"mpal_network_from_ameya.pickle\", \"rb\") as f:\n",
    "        G_mpal = pickle.load(f)\n",
    "    \n",
    "    # reading and processing each vcf file in vcf_files -- getting annotations, files unzipped\n",
    "    processed_vcfs = []\n",
    "    mutationtype_lists = []\n",
    "    v1 = [vcf_files[0]]\n",
    "    for v in v1:\n",
    "        path = f\"/projectnb/paxlab/DATA/ICGC/WGS/LAML-KR/{v}\"\n",
    "        vcf = process_vcf(path, zipped = False, get_annotations = True)\n",
    "        processed_vcfs.append(vcf[0])\n",
    "        mutationtype_lists.append(vcf[1])\n",
    "    print(\"VCF files processed and annotated\")\n",
    "    \n",
    "    # get cumulative mutation type list\n",
    "    for l in range(1, len(mutationtype_lists)):\n",
    "        if mutationtype_lists[0] != mutationtype_lists[l]:\n",
    "            mutationtype_lists[0] = mutationtype_lists[0] + list(set(mutationtype_lists[l]) - set(mutationtype_lists[0]))\n",
    "            \n",
    "    mutation_types = mutationtype_lists[0]\n",
    "    print(f\"all mutation types: {mutation_types}\")\n",
    "    \n",
    "    # get mutations per individual -- all mutation types\n",
    "    mutation_matrix = get_mutation_matrix(processed_vcfs, bins)\n",
    "    print(f\"built mutation matrix -- shape = {mutation_matrix.shape}\")\n",
    "    \n",
    "    # calculate accessibility normalized cell scores for mutation matrix -- all mutation types\n",
    "    cell_scores = get_cell_scores(mutation_matrix, atac_sparsematrix, calculate_total = True)\n",
    "    print(f\"got cell score matrix -- shape = {cell_scores.shape}\")\n",
    "    \n",
    "    # create pagerank dictionaries -- all mutation types\n",
    "    pagerank_dicts = get_pagerank_dicts(cell_scores, atac_cellnames)\n",
    "    print(f\"got {len(pagerank_dicts)} pagerank dictionaries\")\n",
    "    \n",
    "    # get pagerank csv file -- all mutation types\n",
    "    get_pagerank_csv(G_mpal, pagerank_dicts, rna_cellnames, \"/projectnb/paxlab/randers1/pagerank_all.csv\", [\"all1\", \"all2\", \"all3\", \"all4\", \"all5\", \"all6\", \"all7\", \"all8\", \"all\"])\n",
    "    print(\"PageRank is complete, results can be found in the csv file in the directory indicated\")\n",
    "    \n",
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c97cd4ec",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-randerson] *",
   "language": "python",
   "name": "conda-env-.conda-randerson-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
